README.md
# README

![alt text](https://raw.githubusercontent.com/silicator/PiFunk/master/docs/favicon.ico "Logo PiFunk")

## PiFunk Radio Transmitter - with FM/AM-Modulation for HAM-Bands

**Early Experimental! (WIP)**

___

### Configurations

0. Download Linux-Pi-Distro

[Raspbian Buster lastest (Full)](https://downloads.raspberrypi.org/raspbian_full_latest) >= v4.19.66

and extract and load bootable image on SD-card via Rufus

1. Get this program:

`sudo apt-get install git` via git or download it on the page as tar.gz/zip

`git clone https://github.com/silicator/PiFunk`

2. To configure the Pi for modules via menu (I2C, UART etc.): `sudo raspi-config`

Using w1-gpio sometimes needs a 4.7 - 10 kΩ pullup resistor connected on GPIO Pin

(if you have problems deactivate 1-wire config!)

1-Wire by default BCM4 setting needs to be activated in boot-config for autostart

3. Manually open with nano-editor: `sudo nano /boot/config.txt` (i provide one too)

check/add lines:

`dtoverlay=w1-gpio,gpiopin=4,pullup=0` add pullup=1 if needed

`dtoverlay=audio=on` for bcm audio

optional:

`dtparam=spi=on` for SPI support

`dtoverlay=i2c1-bcm2708` for I2C Bus

`enable_uart=1` for UART RX & TX

`init_uart_baud=9600` data transmission rate

`dtoverlay=pps-gpio,gpiopin=18` for GPS-device pps(puls-pro-second)-support

Sync to GPS 1 PPS signal for Pi PCM-Clock (PIN 12 / GPIO 18 = PCM_CLK / PWM0) for accuracy

`sudo nano /etc/modules` opens modules.conf with text editor (provide one too)

`pps-gpio` Add this line

4. Save your changes with ctrl-o <return/enter> then exit with ctrl-x

___

### Installations

5. First update & upgrade system:

`sudo apt-get update` for system updates

`sudo apt-get upgrade` for system upgrades

6. You will need some libraries for this:

a) `sudo apt-get install libraspberrypi-dev raspberrypi-kernel-headers` for Kernel & Firmware

b) `sudo apt-get install libsndfile1-dev` for ALSA SND-lib

or download it directly [SND](https://packages.debian.org/de/sid/libsndfile1-dev)

c) `sudo apt-get install python-dev python3-dev` for py3

d) [RPi.GPIO lib v0.7.0 for Py3](https://files.pythonhosted.org/packages/cb/88/d3817eb11fc77a8d9a63abeab8fe303266b1e3b85e2952238f0da43fed4e/RPi.GPIO-0.7.0.tar.gz) (also in repo)

[RPi.GPIO Project Site](https://pypi.org/project/RPi.GPIO/) 

[RPi.GPIO Sourceforge-Site](https://sourceforge.net/projects/raspberry-gpio-python/files/)

or download via terminal: `sudo wget https://pypi.python.org/packages/source/R/RPi.GPIO/RPi.GPIO-0.7.0.tar.gz`

then extract: `tar -xvf RPi.GPIO-0.7.0.tar.gz`

and install it: `sudo pip-3.7 install RPi.GPIO` for Py3 (easiest way)

or alternative way: `sudo apt-get -y install python3-rpi.gpio`

7. Compiler installation:

a) GNU C Compiler GCC `sudo apt-get install gcc-9.2.0` or g++

b) GNU Debugger GDB `sudo apt-get install gdbserver`

8. `sudo reboot` then reboot to apply the changes

___

### Builds

9. Navigate to directory:

a) `cd PiFunk` with default path: `/home/pi/PiFunk/`

b) GCC Compiler flags:

`-g3` for normal GNU compiler debug informations (1-3 level, 2 is default)

`-ggdb3` for GNU debugger informations level 3

`-Wall` for debug all warning informations

`-Werror` for debug error informations

`-v` Print compilation verbose informations

`-std=c99` (as iso `-std=iso9899:1999` strict)

`-std=gnu99` with additional gnu extension to c99

(`-std=gnu++17` if you like when using g++)

`-pedantic-errors` for error console messages if problem between c99 and gnu extensions

`-Iinclude ` for using include-directory with header files

`-I/opt/vc/include/` for loading bcm header folder

`-L/opt/vc/lib` for loading bcm folder

`-Llib` for using library-directory

`-lm` for math-lib is obligatory!

`-lbcm_host` for loading bcm firmware >= v1.20190718

`-lpthread` lib for process threads

`-lgthread` lib for graphic threads

`-lsndfile` -l links library name for ALSA "snd"-lib

`-shared` for generating shared object libraries

`-c` for compiling without linking for making object

`-E` for stopping after preprocessing stage in compilation

`-D_USE_MATH_DEFINES` for mathematical lib definitions

`-D_GNU_C_SOURCE` for loading GNU C Source Macros for non-std setups

(combining lot of different ones: ISO C89, ISO C99, POSIX.1, POSIX.2, BSD, SVID, X/OPEN, LFS)

`-D_POSIX_C_SOURCE=200809L` for POSIX2 Macros needed with bcm (or 199309L for POSIX1)

`-DRASPI=1` defines the macro to be used by the preprocessor (here the Pi model or 0-4, else std-values 0-3)

 -> will be detected by the Makefile via the type of the ARM-Processor

 (other macros possible if in the C-code implemented)

`-fPIC` for generating position independent code (PIC) for bigger programs

`-O3` for optimization stage 1-3 (memory, speed etc.) via compiler

`-o` for individual output-filename flag

10. Generating libraries:

a) Image of the GCC Flow-diagram for generating [Libraries](docs/GCC_Schema.jpg)

*.c=C-code, *.h=headerfile,

*.i=assembled preprocessor C-code, *.S=assembler-code, *.s=preprocessed assembler-code,

*.o=compiled object, *.lib=library object, *.a=archive object, *.so=shared dynamic library object,

*.out=default binary, pifunk(.bin)=executable binary (program)

b) manually compiling/linking libraries:

`sudo gcc -Wall -Werror -std=gnu99 -pedantic-errors -g3 -ggdb3 -Iinclude -Llib -I/opt/vc/include -lbcm_host -lm -lsndfile -lpthread -shared -O3 -fPIC pifunk.c -D_USE_MATH_DEFINES -D_GNU_SOURCE -DRASPI=1 -o include/pifunk.i lib/pifunk.s lib/pifunk.o lib/pifunk.a lib/pifunk.lib lib/pifunk.so`

c) manually compiling/linking executable binary:

`sudo gcc -Wall -Werror -std=gnu99 -pedantic-errors -g3 -ggdb3 -Iinclude -Llib -I/opt/vc/include -lbcm_host -lm -lsndfile -lpthread -shared -O3 -fPIC pifunk.c -D_USE_MATH_DEFINES -D_GNU_SOURCE -DRASPI=1 -o bin/pifunk`

d) optional Pi-Flags:

 `-march=armv6l` architecture version of ARM ("native" is auto option)

 `-mtune=arm1176jzf-s` special architecture type tuning

 `-mfloat-abi=hard` floating-point ABI to use, permissible values are: ‘soft’, ‘softfp’, ‘hard’

 `-mfpu=vfp` virtual floating point hardware module support, for pi2/3/4 use neon-vfpv4

 `-ffast-math` increase speed for float ops and outside the IEEE-standard and deactivates errno-functions

 `sudo piversion` for checking your pi version

e) Makefile commands:

 `sudo make piversion` for checking your pi version via make

 `sudo make install` for installing pifunk files incl. build folder

 `sudo make uninstall` for uninstalling pifunk files

 `sudo make` for compilation with pre-configured flags for compilation for all Pi's

 `sudo make run` for running with standard pifunk flags

 `sudo make help` for starting help command of pifunk

 `sudo make assistant` for starting step-assistant of pifunk

 `sudo make clean` for removing pifunk.out and pifunk.o files in bin folder if necessary

___

### Preparations

11. Hardware-Setup:

a) Use (original) power supply 10 W, 5 V @ ~2 A or ~5 V/500 mA via mini-USB 2.0 or 5 V Pins possible)

b) Check specifications: my Pi B+ v1.2 @ 700 MHz / 512 MB RAM on ARM processor with driver bcm2835-v1.55

-> SoC from Broadcom	depending on pi model: BCM2711, BCM2835,	BCM2836,	BCM2837,	BCM2837B0

for more infos on other boards just visit [Adafruit](http://www.adafruit.com)

or [Wikipedia Spec Summary](https://de.wikipedia.org/wiki/Raspberry_Pi)

c) Antenna to GPCLK0 (GPIO 4, PIN 7) for PWM (Pulse with Modulation)

@ 2-4 mA (max. 50 mA on ALL PINs and 16 per bank!)

- Antenna should be grounded (see Pinout image) to prevent noise and other problems

- For transmission you should use tested/certified antennas with mounts (BNC/SDA/PL - m/f) if possible

- Tip: You could use just a copper wire for antenna:

 CB 11 m-Band (lambda/2, 5.5 m, 216.535" in) and 70 cm-Band (PMR) (lambda(1/4), 17.0 cm, 6.7" in)

![Pinout](docs/pinout-gpio-pib+.jpg)

d) You can try to smooth the Resistence R out with a 1:X (2-43)-balun if using long HF antenna

- Dummy-load: 1-100 W @ 50 Ohm "cement" or similar (aluminium case) with cooler for testing

e) For handling overheating of the Pi's processor use cooling-ribs with fan (5 V DC/0.2 A - 20x20 mm)

you can overclock the Pi if you want to on own risk but it's not recommended

f) RTC: Module DS3231 uses

3.3 V (PIN 1), SDA0 (PIN 3, GPIO0 on I2C), SCL0 (PIN 5, GPIO1 on I2C) & GND (PIN 9)

-> need to activate I2C in pi config!

![RTC](docs/RTC-top.jpg)

g) GPS Module: Ublox Neo 7M 

Pinout: 5 V (PIN 4), GND (PIN 6), RX to UART-TXD (GPIO 14 PIN 8), TX to UART-RXD (GPIO 15, PIN 10), PPS to PCM_CLK (GPIO 18, PIN 12)

it prints in NMEA format so change config `ttyAMA0` to `tty1`

`sudo cat /dev/ttyAMA0` or alternative `sudo cat /dev/ttyS0`

-> need to activate UART (serial 0) in pi config! Yes here crosswiring!! -> (RX of GPS receives what Pi TX'ed)

![GPS](docs/GPS-Neo7M.jpg)

h) Morse-code-table:

Will be implemented later!

![Morsecode](docs/morsecodeCW.jpg)

___

### Run

12. Run with admin/root permissions:

Arguments: would be best to input in this specific order to prevent problems

Use '. dot' as decimal-comma separator!

`[-n <filename (.wav)>] [ -f <freq (MHz)>] [-s <samplerate (kHz)>] [-m <modulation (fm/am)>] [-c <callsign (optional)>] [-p <power 0-7)>]`

extra single menu-flags: -> no further argument needed

`[-a]` for assistant in step-by-step

`[-h]` for help with more infos and arguments

`[-u]` for extra menu (csv, commandline)

default: `sudo ./pifunk -n sound.wav -f 446.006250 -s 22050 -m fm -c callsign -p 7`

Radio works with .wav-file with 16-bit @ 22050.000 [Hz] mono / 0.1-700 to 1500 MHz range depending on the Pi.

It's recommended not to transmit on frequencies higher than the processor speed at the moment to prevent stuttering/lags,

but results would be interesting to know with overclocking.

Explicit CTSS-Tones (38 included) for PMR can be found here as a list: [CTSS](ctsspmr.csv)

___

### Disclaimers

13. Warnings/Caution:

- Private Project! It's Early Access & Work in Progress (WIP)!

- I'm not a professional so **NO guarantees or warranty** for any damage etc.!!

- Usage at **your own risk** !!

- Check laws of your country first! Some Frequencies & Powerlevels (Watt) are prohibited or need at least a HAM-License!

- Pi operates with square-waves (²/^2)!! Use Low-/High-Band-Pass-Filters with dry (not electrolytic) capacitors

  (C=10-100 pF) with solenoid (ring) toroid (ferrite) chokes (B=10-50 uH like the FT37-43)

  or resistors (R=~10 kOhm), diodes to prevent backflow

  transmission (TX) simultaneously on permitted frequencies! -> [Bandpass-Diagram](docs/filter_600.jpg)

* Help / Testers and Feedback are always appreciated! :)

* Thank you and have fun 73!

___

### Links

14. Additional Guidelines

[GitPage](https://silicator.github.io/PiFunk/)

[Wiki](https://github.com/silicator/PiFunk/wiki)

[Readme Guideline](README.md)

[Contribution Guideline](docs/CONTRIBUTING.md)

[Code of Conduct Guideline](docs/CODE_OF_CONDUCT.md)

[Copying Guideline](docs/COPYING.md)

[License Guideline](LICENSE.md) under Open-Source GPLv3.0

Would appreciate being named in the source, Thank you.

### Acknowledgements

 15. Credits

based on pifm/am, pifmadv scripts/snippets


RPi.GPIO/CHANGELOG.txt
Change Log
==========

0.7.0
-------
- Updated RPI_INFO to include RPi 4B
- Fixed pull up/down for Pi4 (issue 168)
- Fix spelling mistake in docstrings
- Tested and working on Raspbian Buster + Python 3.8.0b2
- Fix board detection for aarch64 (Issues 161 / 165)
- Fix checking mmap return value in c_gpio.c (issue 166)

0.6.5
-----
- Fix exception on re-export of /sys/class/gpio/gpioNN

0.6.4
-----
- Event cleanup bug (issue 145)
- Raise exception for duplicate PWM objects (issue 54 - Thijs Schreijer <thijs@thijsschreijer.nl>)
- Fix build warnings (Issue 146 - Dominik George) 
- test.py runs unchanged for both python 2+3
- Soft PWM stops running fix (Issues 94, 111, 154)
- Soft PWM segfault fix (Luke Allen pull request)

0.6.3
-----
- Fix code so it builds under PyPy (Gasper Zejn)
- os.system breaks event detection - Matt Kimball (issue 127)

0.6.2
-----
- Rewrote Debian packaging mechanism
- RPI_INFO reports Pi 3
- Changed module layout - moved C components to RPi._GPIO

0.6.1
-----
- Update RPI_INFO to detect more board types
- Issue 118 - add_event_detect sometimes gives runtime error with unpriv user 
- Issue 120 - setmode() remembers invalid mode

0.6.0a3
-------
- Now uses /dev/gpiomem if available to avoid being run as root
- Fix warnings with pull up/down on pins 3/5
- Correct base address on Pi 2 when devicetree is disabled
- caddr_t error on compile (Issue 109)
- Error on invalid parameters to setup() (issue 93)
- Add timeout parameter to wait_for_edge() (issue 91)

0.5.11
------
- Fix - pins > 26 missing when using BOARD mode
- Add getmode()
- Raise exception when a mix of modes is used
- GPIO.cleanaup() unsets the current pin mode

0.5.10
------
- Issue 95 - support RPi 2 boards
- Introduce RPI_INFO
- Deprecate RPI_REVISION
- Issue 97 - fixed docstring for setup()

0.5.9
-----
- Issue 87 - warn about pull up/down on i2c pins
- Issue 86/75 - wait_for_edge() bugfix
- Issue 84 - recognise RPi properly when using a custom kernel
- Issue 90 - cleanup() on a list/tuple of channels

0.5.8
-----
- Allow lists/tuples of channels in GPIO.setup()
- GPIO.output() now allows lists/tuples of values
- GPIO.wait_for_edge() bug fixes (issue 78)

0.5.7
-----
- Issue 67 - speed up repeated calls to GPIO.wait_for_event()
- Added bouncetime keyword to GPIO.wait_for_event()
- Added extra edge/interrupt unit tests
- GPIO.wait_for_event() can now be mixed with GPIO.add_event_detect()
- Improved cleanups of events
- Issue 69 resolved

0.5.6
-----
- Issue 68 - support for RPi Model B+
- Fix gpio_function()

0.5.5
-----
- Issue 52 - 'unallocate' a channel
- Issue 35 - use switchbounce with GPIO.event_detected()
- Refactored events code
- Rewrote tests to use unittest mechanism and new test board with loopbacks
- Fixed adding events after a GPIO.cleanup()
- Issue 64 - misleading /dev/mem permissions error
- Issue 59 - name collision with PWM constant and class

0.5.4
-----
- Changed release status (from alpha to full release)
- Warn when GPIO.cleanup() used with nothing to clean up (issue 44)
- Avoid collisions in constants (e.g. HIGH / RISING / PUD_DOWN)
- Accept BOARD numbers in gpio_function (issue 34)
- More return values for gpio_function (INPUT, OUTPUT, SPI, I2C, PWM, SERIAL, UNKNOWN)
- Tidy up docstrings
- Fix /dev/mem access error with gpio_function

0.5.3a
------
- Allow pydoc for non-root users (issue 27)
- Fix add_event_detect error when run as daemon (issue 32)
- Simplified exception types
- Changed from distribute to pip

0.5.2a
------
- Added software PWM (experimental)
- Added switch bounce handling to event callbacks
- Added channel number parameter to event callbacks (issue 31)
- Internal refactoring and code tidy up

0.5.1a
------
- Fixed callbacks for multiple GPIOs (issue 28)

0.5.0a
------
- Added new edge detection events (interrupt handling)
  - Added add_event_detect()
  - Added remove_event_detect()
  - Added add_event_callback()
  - Added wait_for_edge()
- Removed old experimental event functions
  - Removed set_rising_event()
  - Removed set_falling_event()
  - Removed set_high_event()
  - Removed set_low_event()
- Changed event_detected() for new edge detection functionality
- input() now returns 0/LOW == False or 1/HIGH == True (integers) instead of False or True (booleans).
- Fix error on repeated import (issue 3)
- Change SetupException to a RuntimeError so it can be caught on import (issue 25, Chris Hager <chris@linuxuser.at>)
- Improved docstrings of functions

0.4.2a
------
- Fix for installing on Arch Linux (Python 3.3) (issue 20)
- Initial value when setting a channel as an output (issue 19)

0.4.1a
------
- Added VERSION
- Permit input() of channels set as outputs (Eric Ptak <trouch@trouch.com>)

0.4.0a
------
- Added support for Revision 2 boards
- Added RPI_REVISION
- Added cleanup() function and removed automatic reset functionality on program exit
- Added get_function() to read existing GPIO channel functionality (suggestion from Eric Ptak <trouch@trouch.com>)
- Added set_rising_event()
- Added set_falling_event()
- Added set_high_event()
- Added set_low_event()
- Added event_detected()
- Added test/test.py
- Converted debian to armhf
- Fixed C function short_wait() (thanks to Thibault Porteboeuf <thibaultporteboeuf@gmail.com>)

0.3.1a
------
- Fixed critical bug with swapped high/low state on outputs
- Added pull-up / pull-down setup functionality for inputs

0.3.0a
------
- Rewritten as a C extension
- Now uses /dev/mem and SoC registers instead of /sys/class/gpio
- Faster!
- Make call to GPIO.setmode() mandatory
- Added GPIO.HIGH and GPIO.LOW constants

0.2.0
-----
- Changed status from alpha to beta
- Added setmode() to be able to use BCM GPIO 00.nn channel numbers
- Renamed InvalidPinException to InvalidChannelException

0.1.0
------
- Fixed direction bug
- Added MANIFEST.in (to include missing file)
- Changed GPIO channel number to pin number
- Tested and working!

0.0.3a
------
- Added GPIO table
- Refactored
- Fixed a few critical bugs

pifunk.py
#!/usr/bin/python

##PiFunk Radio transmitter for PMR446/CB on FM & AM also for ltp, 433, cb, pmr
##and maybe someday, beacon, gps, internet, relais, aprs, vhf, ts2/3, RDS, morse, echolink
##microphone (usb & jack) + playlist, mp3/wav-Files. (WIP)
## dependency python 3.7.x needeed to run script
##pifm GPIO's: 4 (pin 7 GP-CLK0) and GND (pin 9 = GND) or 14 (pin 8 TXD) & GND (pin 6) & 15 (pin 10 RDX) & 21 (pin 40 SCLK) --> 39 GND
##ARM - Structure on Pi's !!! (can only be emulated on PC, so no real GPIO access!!) my Pi : rev. 2 B+
##------------------------------------------------------------------------------
#Avoid transmitting on 26.995, 27.045, 27.095, 27.145 and 27.195 MHz, as these are Class C radio-control channels,
#and the FCC takes a dim view of voice broadcasts on these frequencies. For that matter,
#re-broadcast of copyrighted material (sports, news and weather programming) is a ##violation of the law,
#and could result in fines, jail time, and confiscation of all radio equipment on your premises.
#UK law is 4 W (4000 mW) / GER 100 mW ERP for PMR and 4 W for CB
#-> sending on square-func means transmission on 3 other freqs so please use a low-pass-filter!
#-------------------------------------------------------------------------------
## py is function-scope not lock-scope!!
## Imports
import io
import os
import sys
import gc
import glob
import string
import StringIO
import struct
import types
import decimal
import array
import argparse
import readline
import rlcomplete
import atexit
import calendar
import time #from time import time
import timeit
import datetime
import site
import re
import math #from math import *
import statistics
import locale
import distutils
import serial
import signal
import asyncio
import tracemalloc
import threading
import builtins
import collections
import queue
import subprocess #from subprocess import run, call, pipe
import multiprocessing
import wave
import random
import logging
import weakref
import socket
import ipaddress
import mimetypes
import urllib.request
## some other plugins
import zipfile
import json
import csv
#import numpy as np
#import scipy -> evtl install
#import scipy.io.wavfile as wavfile
#import matplotlib.pyplot as plt

## RPi & GPIO lib obligatory for controlling gpios
#sudo apt-get install python-rpi.gpio
try:
 import RPi.GPIO as GPIO4
 import RPI.GPIO as GPIO4
except:
 from RPi._GPIO import GPIO4
else:
 from RPi import GPIO4

import gps
import pynmea2 # for gps devices e.g. Neo-7M
#sudo python setup.py install
#hex-code: 0x10A --> dec:26
##------------------------------------------------------------------------------
#loading hardware on startup

def __init__ (self):
  print ("\nWelcome to PiFunk!\n")
  self.data = []
  print (sys.argv)

  try:
          os.system ("sudo modprobe w1-gpio") #rpi 1-2
  except:
          os.system ("sudo rmmod w1-gpio") # rpi3

  base_dir = "/sys/bus/w1/devices/"
  device_folder = glob.glob (base_dir + "28*") [0]
  device_file = device_folder + "/w1_slave"

  cpid = os.fork ()
  if not cpid:
      os._exit (0)
      os.waitpid (cpid, 0)

  pullup = 0
  DEBUG = 1
  LOGGER = 1
  gpio_pin = 4
  GPIO.initialize ()
  GPIO.setwarnings (False)
  GPIO.setmode (GPIO.BCM)
  GPIO.setmode (GPIO.BOARD)
  GPIO.setup (4, GPIO.OUT) #or 11 if used
  output = GPIO.output (4, TRUE)
  sensor_data = (gpio_pin, GPIO.PINS.GND, GPIO.PINS.RXD, GPIO.PINS.TXD)
  pi_pwm = GPIO.PWM (4, freq)
  pi_pwm.start (0)

  while True:
      for Duty in range (0, 60, 1)
      pi_pwm.ChangeDutyCycle (Duty)
##------------------------------------------------------------------------------

def soundfile (filename):
  filename = char (input ("\nEnter filename (*.wav): "))
  if filename != 0:
      print ("\nFilename is: " + filename + " \n")
      return filename
  else:
      print ("\nNo custom filename specified! using standard file sound.wav !! \n")
      filename = char ("sound.wav")
      return filename

def frequency (freq):
  freq = float (input ("\nEnter frequency (1 kHz-1000 MHz) with . as decimal (5 digit accuracy):  "))
  for samplerate in range (0.000001, 1000)
  if freq > 0:
      print ("\nFrequency is: " + freq " MHz \n")
      return freq
  else:
      freq = float (446.00000)
      print ("\nFrequency must be > 0 !! Using 446.00000 MHz instead! \n")
      return freq

def sampler (samplerate):
  samplerate = int (input ("\nEnter samplerate (22050/44100/48000 kHz): "))
  if samplerate > 0:
      print ("\nSamplerate is: " + samplerate + " \n")
      return samplerate
  else:
      samplerate = int (22050)
      print ("\nNo custom samplerate specified or negative! Using standard 22050 kHz!! \n")
      return samplerate

def channels (channels):
  channels = int (input ("\nEnter number of channels (1 mono or 2 stereo): "))
  if channels 1 || 2:
      print ("\nChannels: " + channels + " \n")
      return channels
  else:
      channels = 1
      print ("\nNo channels specified! Using standard: 1 for mono \n")
      return channels

def modulation (mod):
  mod = char (input ("\nEnter modulation type (fm/am): ")).lower()
  if mod == "fm" || "am":
      print ("\nModulation is: " + mod + " \n")
      return mod
  else:
      mod = char ("fm")
      print ("\nNo mode specified! Using standard fm \n")
      return mod

def callname (callsign):

  callsign = char (input ("\nEnter callsign: "))
  if callsign != 0:
      print ("\nCallsign is: " + callsign + " \n")
      return callsign
  else:
      callsign = char ("callsign")
      print ("\nNo callsign specified! Using standard *callsign* \n")
      return callsign

def c_arg_parser ():
  print ("\nParsing args to C ... \n")
  return filename, freq, samplerate, mod, callsign, power

def play_wav ():
  print ("\nPlaying wav ... \n")
  if (filename && freq && samplerate && mod != None):
      call (["sudo ./pifunk ", filename, freq, samplerate, mod, callsign])
  else:
      call (["sudo ./pifunk sound.wav 100.00000 22050 fm callsign"])

def play_mp3 ():
  print ("\nPiping mp3 to ffmpeg ... \n")
  ffmpeg -i sounds/sound.mp3 -f s16le -ar 22.05k -ac 1 | sudo ./pifunk -100.00000 fm

def play_mic ():
  ## Broadcast from a (usb) microphone, stereo
  print ("\nUsing mic and piping through aracord ... \n")
  arecord -d0 -c2 -f S16_LE -r 22050 -twav -D copy | sudo ./pifunk 100.00000 fm
  arecord -D plughw:1,0 -c1 -d 0 -r 22050 -f S16_LE | sudo ./pifunk -f 100.00000 fm

def play_net ():
  $port = 80 #http web
  card = 0 # microphone devices
  subdevice = 0
  print ("\nUsing network stream and piping through arecord ... \n")
  arecord -D hw:${card},${subdevice} -f S16_LE -r 22050 -t wav | sudo nc -1 ./pifunk 100.0000 $port
  arecord -D hw:${0},${0} -f S16_LE -r 22050 -t wav | sudo nc -1 ./pifunk 100.00000 $port

def led_on (pin):
  GPIO.output (pin, GPIO.HIGH)
  #time.sleep(1)
  print ("\nLED on \n")
  return pin

def led_off (pin):
  GPIO.output (pin, GPIO.LOW)
  #time.sleep(1)
  print ("\nLED off \n")
  return pin

def led_blink (pin):
  led_on (pin)
  time.sleep (1)
  led_off (pin)
  time.sleep (1)

def led_blinking (pin):
  print ("\nBlinking 1/sec \n")
  for i in range (0, 61, 1): led_blink (11)
  GPIO.cleanup ()

def csv_reader ():
  csv_file_object = csv.reader (open ("docs/ctsspmr.csv", "rb"))
  print ("\nImporting CTSS-Table \n")
  with open ("docs/ctsspmr.csv", "rb") as f:
  reader = csv.reader (f, delimiter = ",", quotechar = "|")
  for row in reader: print (",".join (row))
  # reading single infos of ctss tones later (wip)
  print ("\nImporting completed \n")
  return reader

def logger ():
  with open ("logs/log.txt", "w") as f:
  call (["sudo python ", "./pifunk-log.py"], stdout = f)
  print ("\nLogging ... \n")


if __name__ == "__main__":
  print ("\nStarting Program!\n")
  (int (sys.argv [1])
##------------------------------------------------------------------------------
## run another py-script from shell-terminal (holds main script)
## selecting a individual band:

#subprocess.run (["sudo", "python", "pi-gpio.py"])

#subprocess.run (["sudo", "python", "pifunk-pmr.py"])

#subprocess.run (['sudo', 'python', 'pifunk-cb.py'])

#subprocess.run (["sudo", "python", "pifunk-temp.py"])

#subprocess.run (["sudo", "python", "pifunk-cpu-ram-disk.py"])

#subprocess.run (["sudo", "python", "pi-minidisplay.py"])

## run shell/bat-script if neccessary

##------------------------------------------------------------------------------
# main program
__init__ ()
datetime.now().strftime("%d-%m-%Y, %H:%M:%S \n")
#print (datetime.datetime.now ())
#current_time = datetime.datetime.now ()
#str (datetime.now ())
#current_time.isoformat ()

#here a menu with switchcase
logger ()
csv_reader ()

#if args not specified, else so to play
soundfile ()
frequency ()
sampler ()
channels ()
modulation ()
callsign ()

c_arg_parser ()

try:
#if all args are parsed so to transmission mode
 led_on ()
 led_blinking ()
 led_off ()

#here a switchcase later
 play_wav ()
 play_mp3 ()
 play_net ()
 play_mic ()

except KeyboardInterrupt:
  GPIO.cleanup()       # clean up GPIO on CTRL+C exit

GPIO.cleanup()           # clean up GPIO on normal exit
##------------------------------------------------------------------------------
##test-area
#nosetests
print ("\nEnd of script!\n")
